Harden 'search' command against malformed assets

### What you were trying to do

Search for stuff in the in-game console, like `search stone`

### What actually happened

Got a one-line error in the game console. Turned out an asset was misconfigured (was pointing to a template asset in the wrong module). After fixing that another similar case hit where the JSON in another asset was malformed, yet the error didn't give any clue _which _ asset.

Turns out I guess that to be able to search the engine goes and looks at every possible asset which means if it encounters even a single broken asset it'll error out and the search fails.

#1402 is a bigger goal to better validate assets (warnings at compile time somehow? Code scanning step?). But maybe we can harden `search` to better log what is breaking it?

#2874 also seems mildly similar

### How to reproduce

* Break an asset (like remove a comma or mess up a module name in a URI)
* Run the game and make sure the asset is within the active module set
* Run a `search` command for anything - get error!

### Example log 

`An error occurred while executing command 'search': Unable to resolve based block definition 'CoreBlocks:rock'`

(in that case the `rock` template actually lived in the CoreAssets module instead)

I think I hit a different variant of this or just didn't record a more full log snippet. This one is hair-pulling levels of frustration as it points out _exactly where in a file the error is_ .... but not in _which_ file / prefab 😛 

This is with the entirety of JoshariasSurvival enabled as of this writing, but maybe I'll find it and fix it before long ...

```
21:23:12.399 [main] ERROR o.t.logic.console.ConsoleImpl - An error occurred while executing a command
org.terasology.logic.console.commandSystem.exceptions.CommandExecutionException: com.google.gson.JsonSyntaxException: com.google.gson.stream.MalformedJsonException: Expected name at line 12 column 6 path $.entity.prefab
        at org.terasology.logic.console.commandSystem.AbstractCommand.execute(AbstractCommand.java:242)
        at org.terasology.logic.console.ConsoleImpl.execute(ConsoleImpl.java:264)
        at org.terasology.logic.console.ConsoleImpl.execute(ConsoleImpl.java:228)
        at org.terasology.logic.console.ui.ConsoleScreen.lambda$initialise$0(ConsoleScreen.java:79)
        at org.terasology.rendering.nui.widgets.UIText.onKeyEvent(UIText.java:468)
        at org.terasology.logic.console.ui.UICommandEntry.onKeyEvent(UICommandEntry.java:83)
        at org.terasology.rendering.nui.internal.NUIManagerInternal.keyEvent(NUIManagerInternal.java:648)
        at org.terasology.rendering.nui.internal.NUIManagerInternalMethodAccess.invoke(Unknown Source)
        at org.terasology.entitySystem.event.internal.EventSystemImpl$ByteCodeEventHandlerInfo.invoke(EventSystemImpl.java:531)
        at org.terasology.entitySystem.event.internal.EventSystemImpl.sendConsumableEvent(EventSystemImpl.java:301)
        at org.terasology.entitySystem.event.internal.EventSystemImpl.send(EventSystemImpl.java:280)
        at org.terasology.entitySystem.entity.internal.BaseEntityRef.send(BaseEntityRef.java:205)
        at org.terasology.input.InputSystem.send(InputSystem.java:503)
        at org.terasology.input.InputSystem.sendKeyEvent(InputSystem.java:447)
        at org.terasology.input.InputSystem.processKeyboardInput(InputSystem.java:382)
        at org.terasology.input.InputSystem.update(InputSystem.java:135)
        at org.terasology.engine.modes.StateIngame.handleInput(StateIngame.java:192)
        at org.terasology.engine.subsystem.lwjgl.LwjglInput.postUpdate(LwjglInput.java:55)
        at org.terasology.engine.TerasologyEngine.tick(TerasologyEngine.java:471)
        at org.terasology.engine.TerasologyEngine.mainLoop(TerasologyEngine.java:425)
        at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:401)
        at org.terasology.engine.Terasology.main(Terasology.java:162)
Caused by: com.google.gson.JsonSyntaxException: com.google.gson.stream.MalformedJsonException: Expected name at line 12 column 6 path $.entity.prefab
        at com.google.gson.internal.Streams.parse(Streams.java:60)
        at com.google.gson.TreeTypeAdapter.read(TreeTypeAdapter.java:54)
        at com.google.gson.Gson.fromJson(Gson.java:879)
        at com.google.gson.Gson.fromJson(Gson.java:817)
        at org.terasology.world.block.loader.BlockFamilyDefinitionFormat.load(BlockFamilyDefinitionFormat.java:93)
        at org.terasology.world.block.loader.BlockFamilyDefinitionFormat.load(BlockFamilyDefinitionFormat.java:70)
        at org.terasology.assets.module.UnloadedAssetData$AssetSourceResolver.load(UnloadedAssetData.java:292)
        at org.terasology.assets.module.UnloadedAssetData.load(UnloadedAssetData.java:180)
        at org.terasology.assets.module.ModuleAssetDataProducer.getAssetData(ModuleAssetDataProducer.java:228)
        at org.terasology.assets.AssetType.lambda$reload$2(AssetType.java:359)
        at java.security.AccessController.doPrivileged(Native Method)
        at org.terasology.assets.AssetType.reload(AssetType.java:357)
        at org.terasology.assets.AssetType.getNormalAsset(AssetType.java:386)
        at org.terasology.assets.AssetType.getAsset(AssetType.java:262)
        at org.terasology.assets.management.AssetManager.getAsset(AssetManager.java:223)
        at org.terasology.logic.console.commands.CoreCommands.lambda$findBlockMatches$6(CoreCommands.java:257)
        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
        at java.util.Iterator.forEachRemaining(Iterator.java:116)
        at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
        at org.terasology.logic.console.commands.CoreCommands.findBlockMatches(CoreCommands.java:259)
        at org.terasology.logic.console.commands.CoreCommands.search(CoreCommands.java:199)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.terasology.logic.console.commandSystem.AbstractCommand.execute(AbstractCommand.java:237)
        ... 21 common frames omitted
Caused by: com.google.gson.stream.MalformedJsonException: Expected name at line 12 column 6 path $.entity.prefab
        at com.google.gson.stream.JsonReader.syntaxError(JsonReader.java:1567)
        at com.google.gson.stream.JsonReader.doPeek(JsonReader.java:509)
        at com.google.gson.stream.JsonReader.hasNext(JsonReader.java:418)
        at com.google.gson.internal.bind.TypeAdapters$29.read(TypeAdapters.java:742)
        at com.google.gson.internal.bind.TypeAdapters$29.read(TypeAdapters.java:743)
        at com.google.gson.internal.bind.TypeAdapters$29.read(TypeAdapters.java:718)
        at com.google.gson.internal.Streams.parse(Streams.java:48)
        ... 51 common frames omitted
21:23:12.399 [main] INFO  o.t.logic.console.ConsoleImpl - [ERROR] An error occurred while executing command 'search': com.google.gson.stream.MalformedJsonException: Expected name at line 12 column 6 path $.entity.prefab
```

--- a/engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java
+++ b/engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java
@@ -267,11 +267,9 @@ public class CoreCommands extends BaseComponentSystem {
             }
             // If a prefab is broken , it will throw an exception
             catch(Exception e){
-                List<String> EmptyList=new ArrayList<String>();
                 console.addMessage("Note : Search may not return results if invalid assets are present");
                 console.addMessage("Error parsing : "+curUrn.toString());
                 console.addMessage(e.toString());
-                // return(EmptyList);
                 continue;
             }
         }

--- a/engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java
+++ b/engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java
@@ -49,6 +49,7 @@ import org.terasology.logic.console.suggesters.CommandNameSuggester;
 import org.terasology.logic.console.suggesters.ScreenSuggester;
 import org.terasology.logic.console.suggesters.SkinSuggester;
 import org.terasology.logic.inventory.events.DropItemEvent;
+import org.terasology.logic.location.Location;
 import org.terasology.logic.location.LocationComponent;
 import org.terasology.logic.permission.PermissionManager;
 import org.terasology.math.Direction;
@@ -83,15 +84,10 @@ import org.terasology.world.block.loader.BlockFamilyDefinition;
 
 import java.io.IOException;
 import java.net.UnknownHostException;
+import java.util.*;
 import java.util.concurrent.Callable;
 import java.util.stream.Collectors;
 import java.util.stream.StreamSupport;
-import java.util.List;
-import java.util.Set;
-import java.util.LinkedList;
-import java.util.Locale;
-import java.util.Optional;
-import java.util.Arrays;
 
 /**
  * Adds a series of useful commands to the game. Likely these could be moved to more fitting places over time.
@@ -253,10 +249,29 @@ public class CoreCommands extends BaseComponentSystem {
      * @return List of blocks that match searched string
      */
     private List<String> findBlockMatches(String searchLowercase) {
-        return assetManager.getAvailableAssets(BlockFamilyDefinition.class)
-                .stream().<Optional<BlockFamilyDefinition>>map(urn -> assetManager.getAsset(urn, BlockFamilyDefinition.class))
-                .filter(def -> def.isPresent() && def.get().isLoadable() && matchesSearch(searchLowercase, def.get()))
-                .map(r -> new BlockUri(r.get().getUrn()).toString()).collect(Collectors.toList());
+        ResourceUrn curUrn;
+        List<String> l=new ArrayList<String>();
+        for (ResourceUrn urn : assetManager.getAvailableAssets(BlockFamilyDefinition.class)) {
+            // Current urn for logging purposes to find the broken urn
+            curUrn = urn;
+            try{
+                Optional<BlockFamilyDefinition> def = assetManager.getAsset(urn, BlockFamilyDefinition.class);
+                if (def.isPresent() && def.get().isLoadable() && matchesSearch(searchLowercase, def.get())) {
+                    String s = new BlockUri(def.get().getUrn()).toString();
+                    l.add(s);
+                }
+            }
+            // If a prefab is broken , it will throw an exception
+            catch(Exception e){
+                List<String> EmptyList=new ArrayList<String>();
+                console.addMessage("Note : Search may not return results if invalid assets are present");
+                console.addMessage("Error parsing : "+curUrn.toString());
+                console.addMessage(e.toString());
+                return(EmptyList);
+            }
+        }
+        return l;
+
     }
 
     /**
@@ -523,17 +538,16 @@ public class CoreCommands extends BaseComponentSystem {
         } else {
             dir.set(Direction.FORWARD.getVector3f());
         }
-        Quat4f rotation = Quat4f.shortestArcQuat(Direction.FORWARD.getVector3f(), dir);
 
-        Optional<Prefab> prefab = Assets.getPrefab(prefabName);
-        if (prefab.isPresent() && prefab.get().getComponent(LocationComponent.class) != null) {
-            entityManager.create(prefab.get(), spawnPos, rotation);
-            return "Done";
-        } else if (!prefab.isPresent()) {
-            return "Unknown prefab";
-        } else {
-            return "Prefab cannot be spawned (no location component)";
-        }
+        return Assets.getPrefab(prefabName).map(prefab -> {
+            LocationComponent loc = prefab.getComponent(LocationComponent.class);
+            if (loc != null) {
+                entityManager.create(prefab, spawnPos);
+                return "Done";
+            } else {
+                return "Prefab cannot be spawned (no location component)";
+            }
+        }).orElse("Unknown prefab");
     }
 
     /**

--- a/engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java
+++ b/engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java
@@ -84,7 +84,12 @@ import org.terasology.world.block.loader.BlockFamilyDefinition;
 
 import java.io.IOException;
 import java.net.UnknownHostException;
-import java.util.*;
+import java.util.List;
+import java.util.Set;
+import java.util.ArrayList;
+import java.util.Locale;
+import java.util.Optional;
+import java.util.Arrays;
 import java.util.concurrent.Callable;
 import java.util.stream.Collectors;
 import java.util.stream.StreamSupport;
@@ -250,15 +255,14 @@ public class CoreCommands extends BaseComponentSystem {
      */
     private List<String> findBlockMatches(String searchLowercase) {
         ResourceUrn curUrn;
-        List<String> l=new ArrayList<String>();
+        List<String> outputList=new ArrayList<String>();
         for (ResourceUrn urn : assetManager.getAvailableAssets(BlockFamilyDefinition.class)) {
             // Current urn for logging purposes to find the broken urn
             curUrn = urn;
             try{
                 Optional<BlockFamilyDefinition> def = assetManager.getAsset(urn, BlockFamilyDefinition.class);
                 if (def.isPresent() && def.get().isLoadable() && matchesSearch(searchLowercase, def.get())) {
-                    String s = new BlockUri(def.get().getUrn()).toString();
-                    l.add(s);
+                    outputList.add(new BlockUri(def.get().getUrn()).toString());
                 }
             }
             // If a prefab is broken , it will throw an exception
@@ -267,10 +271,11 @@ public class CoreCommands extends BaseComponentSystem {
                 console.addMessage("Note : Search may not return results if invalid assets are present");
                 console.addMessage("Error parsing : "+curUrn.toString());
                 console.addMessage(e.toString());
-                return(EmptyList);
+                // return(EmptyList);
+                continue;
             }
         }
-        return l;
+        return outputList;
 
     }
 
@@ -538,16 +543,17 @@ public class CoreCommands extends BaseComponentSystem {
         } else {
             dir.set(Direction.FORWARD.getVector3f());
         }
+        Quat4f rotation = Quat4f.shortestArcQuat(Direction.FORWARD.getVector3f(), dir);
 
-        return Assets.getPrefab(prefabName).map(prefab -> {
-            LocationComponent loc = prefab.getComponent(LocationComponent.class);
-            if (loc != null) {
-                entityManager.create(prefab, spawnPos);
-                return "Done";
-            } else {
-                return "Prefab cannot be spawned (no location component)";
-            }
-        }).orElse("Unknown prefab");
+        Optional<Prefab> prefab = Assets.getPrefab(prefabName);
+        if (prefab.isPresent() && prefab.get().getComponent(LocationComponent.class) != null) {
+            entityManager.create(prefab.get(), spawnPos, rotation);
+            return "Done";
+        } else if (!prefab.isPresent()) {
+            return "Unknown prefab";
+        } else {
+            return "Prefab cannot be spawned (no location component)";
+        }
     }
 
     /**

